### ISSUE TITLE
Need advice before back porting some fixes

### ISSUE THREAD

**Issue by samstickland** [attr-encrypted/attr_encrypted#189]
Hi,

I'm in the process of trying to upgrade a very large Rails app from 2.3.18 / Ruby 1.8.7 to at least something approaching modern.

For reasons I won't bore you with, we need to move to Ruby 1.9.3 before moving to Rails 3.

I have a problem with attr_encrypted. The last version that seems to support Rails 2 has a bug with Ruby 1.9.3 and UTF-8 handling, where - in UTF-8 'mode' - the encrypted attributes written to the database are blank.

This is fixed in later versions, but of course, these versions won't run on Rails 2 (error: uninitialized constant AttrEncrypted::Encryptor (NameError)).

So it seems like at this stage our only option is to fork attr_encrypted 1.0.9 and fix the UTF-8 issues in this release.

Can anyone point me towards the code or PR that fixed UTF-8?

Thanks!

Sam

**Comment by samstickland** [comment]
Hi,

Just to add some more info. I've found this issue:

https://github.com/attr-encrypted/attr_encrypted/issues/35

But the PR referenced seems to reference a dead github repository

https://github.com/yellow-lab/attr_encrypted_1_9_3_charset_fix

**Comment by saghaulor** [comment]
@samstickland can you provide more info like part of your stacktrace? I'd like to know where in attr_encrypted this is happening. I have an idea on how to get around this but I need to know more.

**Comment by samstickland** [comment]
Thanks,

Do you mean the stack trace for "error: uninitialized constant AttrEncrypted::Encryptor (NameError)" or related to receiving nil for the attribute value?

**Comment by saghaulor** [comment]
@samstickland the one related to `uninitialized constant`. Let's see if we can get you moving forward without having to fork and backport.

**Comment by samstickland** [comment]
@saghaulor Thanks, not having to fork and backport would be awesome :)

This is the full backtrace from a small TestUnit test I created to test the attr_encrypted functionality.

```
NOTE: Gem.source_index is deprecated, use Specification. It will be removed on or after 2011-11-01.
Gem.source_index called from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/rails-2.3.18/lib/rails/gem_dependency.rb:21.
/home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/inflector.rb:3:in `<top (required)>': iconv will be deprecated in the future, use String#encode instead.
DEPRECATION WARNING: metaclass is deprecated and will be removed from Rails 2.3 (use singleton_class instead). (called from included at /home/sam/XXX/XXX/vendor/plugins/authlogic/lib/authlogic/session/callbacks.rb:70)
DEPRECATION WARNING: ActiveSupport::Dependencies.load_paths is deprecated, please use autoload_paths instead. (called from block in evaluate_init_rb at /home/sam/XXX/XXX/vendor/plugins/fitter_happier/init.rb:6)
/home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/dependencies.rb:466:in `rescue in load_missing_constant': uninitialized constant AttrEncrypted::Encryptor (NameError)
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/dependencies.rb:462:in `load_missing_constant'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/dependencies.rb:106:in `const_missing_with_dependencies'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/rake-0.8.7/lib/rake.rb:2503:in `const_missing'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/aws-s3-0.6.3/lib/aws/s3/extensions.rb:212:in `const_missing_from_s3_library'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/attr_encrypted-1.3.5/lib/attr_encrypted.rb:117:in `attr_encrypted'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/attr_encrypted-1.3.5/lib/attr_encrypted/adapters/active_record.rb:51:in `attr_encrypted'
	from /home/sam/XXX/XXX/app/models/letter.rb:54:in `<class:Letter>'
	from /home/sam/XXX/XXX/app/models/letter.rb:2:in `<top (required)>'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/polyglot-0.3.3/lib/polyglot.rb:63:in `require'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/polyglot-0.3.3/lib/polyglot.rb:63:in `require'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/dependencies.rb:184:in `require'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/dependencies.rb:291:in `require_or_load'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/dependencies.rb:451:in `load_missing_constant'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/dependencies.rb:106:in `const_missing_with_dependencies'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/rake-0.8.7/lib/rake.rb:2503:in `const_missing'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/aws-s3-0.6.3/lib/aws/s3/extensions.rb:212:in `const_missing_from_s3_library'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/dependencies.rb:118:in `const_missing'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/inflector.rb:375:in `const_get'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/inflector.rb:375:in `block in constantize'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/inflector.rb:374:in `each'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/inflector.rb:374:in `constantize'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/core_ext/string/inflections.rb:162:in `constantize'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activerecord-2.3.18/lib/active_record/observer.rb:149:in `block in observe'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activerecord-2.3.18/lib/active_record/observer.rb:149:in `collect!'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activerecord-2.3.18/lib/active_record/observer.rb:149:in `observe'
	from /home/sam/XXX/XXX/app/models/supporter_actions_observer.rb:3:in `<class:SupporterActionsObserver>'
	from /home/sam/XXX/XXX/app/models/supporter_actions_observer.rb:2:in `<top (required)>'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/polyglot-0.3.3/lib/polyglot.rb:63:in `require'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/polyglot-0.3.3/lib/polyglot.rb:63:in `require'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/dependencies.rb:184:in `require'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/dependencies.rb:291:in `require_or_load'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/dependencies.rb:451:in `load_missing_constant'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/dependencies.rb:106:in `const_missing_with_dependencies'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/rake-0.8.7/lib/rake.rb:2503:in `const_missing'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/aws-s3-0.6.3/lib/aws/s3/extensions.rb:212:in `const_missing_from_s3_library'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/dependencies.rb:118:in `const_missing'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/inflector.rb:375:in `const_get'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/inflector.rb:375:in `block in constantize'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/inflector.rb:374:in `each'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/inflector.rb:374:in `constantize'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/core_ext/string/inflections.rb:162:in `constantize'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activerecord-2.3.18/lib/active_record/observer.rb:38:in `block in instantiate_observers'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activerecord-2.3.18/lib/active_record/observer.rb:36:in `each'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activerecord-2.3.18/lib/active_record/observer.rb:36:in `instantiate_observers'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/actionpack-2.3.18/lib/action_controller/dispatcher.rb:15:in `block in define_dispatcher_callbacks'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/callbacks.rb:182:in `call'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/callbacks.rb:182:in `evaluate_method'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/callbacks.rb:166:in `call'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/callbacks.rb:90:in `block in run'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/callbacks.rb:90:in `each'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/callbacks.rb:90:in `run'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/activesupport-2.3.18/lib/active_support/callbacks.rb:276:in `run_callbacks'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/actionpack-2.3.18/lib/action_controller/dispatcher.rb:51:in `run_prepare_callbacks'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/rails-2.3.18/lib/initializer.rb:631:in `prepare_dispatcher'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/rails-2.3.18/lib/initializer.rb:185:in `process'
	from /home/sam/.rvm/gems/ruby-1.9.3-p551/gems/rails-2.3.18/lib/initializer.rb:113:in `run'
	from /home/sam/XXX/XXX/config/environment.rb:28:in `<top (required)>'
	from /home/sam/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/site_ruby/1.9.1/rubygems/custom_require.rb:36:in `require'
	from /home/sam/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/site_ruby/1.9.1/rubygems/custom_require.rb:36:in `require'
	from /home/sam/XXX/XXX/test/test_helper.rb:3:in `<top (required)>'
	from /home/sam/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/site_ruby/1.9.1/rubygems/custom_require.rb:36:in `require'
	from /home/sam/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/site_ruby/1.9.1/rubygems/custom_require.rb:36:in `require'
	from test/unit/supporter_test.rb:2:in `<main>'
```

**Comment by samstickland** [comment]
Well, I've been investigating a little more and it gets weirder.

In the Gem when these lines are present all the encrypted columns are read/written as null:

      attr_reader encrypted_attribute_name.to_sym unless instance_methods.include?(encrypted_attribute_name)                                                                            
      attr_writer encrypted_attribute_name.to_sym unless instance_methods.include?("#{encrypted_attribute_name}=")                                                                      

I don't see how merely defining attr_reader and attr_writer could make this happen, but it does.

With attr_reader and attr_writer defined by the Gem:

    Supporter.find(3).encrypted_first_name
    => null

With attr_reader and attr_writer removed from the gem:

    Supporter.find(3).encrypted_first_name
    => "tTzb40DTDkN+XnOH8lLxUg==\n"

Stranger than that, after removing the reader and writers so the retrieved values are non-null, I get start getting this error on decryption: OpenSSL::Cipher::CipherError: wrong final block length, which is similar to this unsolved issue here: https://github.com/attr-encrypted/attr_encrypted/issues/44

But I've checked the and the encrypted value stored in a new object is identical to that fetched from the database:

```
s = Supporter.build(:first_name => 'Bob', ....)
s.save!
=> true
Supporter.find(s.id).encrypted_first_name == s.encrypted_first_name
=> true
Supporter.find(s.id).first_name
=> OpenSSL::Cipher::CipherError: wrong final block length
```

I'm dropping back to Ruby 1.8.7 just to confirm how things were working there.

**Comment by samstickland** [comment]
OK, so just answering myself here.. The attr_reader and attr_writer are (obviously) over-riding the instance methods that read/write to the DB, and the code that's then supposed to pull this back in is presumably in adapters/active_record.rb..

Still digging around.. although, I don't understand why there is a need to overwrite the getters and setters for the encrypted attribute column? The setter I presume is for safety, but the getter? Why does that exist?

**Comment by samstickland** [comment]
So on 1.8.7 with the same versions of the gem everything works.

   Supporter.find(3).encrypted_first_name
   => "tTzb40DTDkN+XnOH8lLxUg==\n"
   Supporter.find(3).first_name
   => "Daniel"

Note that the encrypted_first_name value is the same as the on 1.9.3  but it is decrypted correctly now.

**Comment by samstickland** [comment]
Right, I've found the problems with getting version 1.0.9 to run on Ruby 1.9.3.

It's in two subtle changes to the language.

instance_methods returns strings in 1.8.7 and symbols in 1.9.3
to_s called on an array containing one string returns the string in 1.8.7, but in 1.9.3 it returns a string surrounded with [ and ]

I've fixed those errors in a fork of 1.0.9 and now we can successfully read and save encrypted values on Ruby 1.9.3 and Rails 2.

The whole UTF-8 thing was massive red herring!

**Comment by saghaulor** [comment]
@samstickland I'm glad that you found a solution.

Do you mind linking your fork so that I can see what you did to resolve the problem? Thanks.

**Comment by samstickland** [comment]
@saghaulor 

Sorry about the late reply. I'm upgrading this application as volunteer work for a charity and I typically only work on it one day a week.

This is the commit in my fork of 1.0.9 with fixes for Ruby 1.9.3:

https://github.com/samstickland/attr_encrypted_fork/commit/af84538c2221bc5936755da8aef2e633d209b97c

Pretty straight forwards once you know were to look!

Note that my change for searching instance_methods will break on Ruby 1.8.7, since this version of ruby returns strings and I search for symbol matches.

Both of these fixes are in later versions of attr_encrypted including a more generic fix for searching instance methods (see https://github.com/attr-encrypted/attr_encrypted/blob/master/lib/attr_encrypted.rb#L435)

In my fork I changed the name of the repository and then bumped the version in the gemspec from 1.0.9 to 1.0.9.1. I wasn't sure how best to handle a fork. If I change the name of the gem it could affect other gems with dependencies on attr_encrypted... unless there's a way to specify an 'alias' for a gem?
